### PR TITLE
tt: initial support for Tarantool 3.0 instances.

### LINKED ISSUE #505: Add support tarantool 3.0
The algorithm for starting tarantool 3.0 differs from the algorithm for starting tarantool 1.10 to 2.11. It now uses a config (local or stored in etcd) to start. In scope of this task basic commands to manage instance like `start` / `stop` / `status` / `restart` / ... must be maintained for tarantool 3.0 .

### PR DESCRIPTION
**Pull request by psergee**
Added initial support for starting Tarantool 3.0 instances using cluster config.
`setup-tarantool` github action does not install 3.0 build yet, so CI is not updated.

Closes #505

### CODE REVIEW

File: cli/cmd/start.go [hunk 1]
```diff
@@ -48,45 +49,64 @@ func NewStartCmd() *cobra.Command {
...
+	return nil
+}
+
 // internalStartModule is a default start module.
 func internalStartModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
 	if !isConfigExist(cmdCtx) {
 		return errNoConfig
 	}
 
+	if cmdCtx.Cli.TarantoolCli.Executable == "" {
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/cmd/start.go:92]: There is no need for the start command to collect apps and instances info if there is no tarantool executable. If tarantool presents - we collect instances info, and perform other checks to make sure that we are able to start found instances. Yes, we can move a check for binary presence after `FillCtx` and this will work, but this is not optimal from my point of view.

File: cli/running/running.go [hunk 2]
```diff
@@ -271,40 +382,20 @@ func CollectInstances(appName string, appDir string) ([]InstanceCtx, error) {
...
 	// If appName equals to base directory name, current working
 	// directory is considered as application to work with.
-	dirPath, luaPath := "", ""
-	if filepath.Base(appDir) == appName {
-		dirPath = appDir
-	} else {
-		luaPath = filepath.Join(appDir, appName+".lua")
-		dirPath = filepath.Join(appDir, appName)
+	if luaPath, err := searchApplicationScript(applicationsDir, appName); err != nil {
+		return []InstanceCtx{}, nil
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/running/running.go]: By mistake. Fixed. Unit test is added.

File: cli/running/running.go [hunk 3]
```diff
@@ -330,6 +425,146 @@ func createLogger(run *InstanceCtx) *ttlog.Logger {
...
+
+		for _, inst := range collectedInstances {
+			var instance = inst
+
+			if ttAppOpts != nil {
+				instance.LogMaxSize = ttAppOpts.LogMaxSize
+				instance.LogMaxAge = ttAppOpts.LogMaxAge
+				instance.LogMaxBackups = ttAppOpts.LogMaxBackups
+				instance.Restartable = ttAppOpts.Restartable
+				setInstCtxFromTtConfig(&instance, ttAppOpts, ttConfigDir)
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/running/running.go]: Moved.

File: cli/running/running.go [hunk 4]
```diff
@@ -528,3 +700,23 @@ func GetAppInstanceName(instance InstanceCtx) string {
 	}
 	return fullInstanceName
 }
+
+// isAbleToStartInstances checks if it is possible to start instances.
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/running/running.go]: Fixed.

**Review by askalt — APPROVED** [review]
lgtm

File: cli/cluster/errors.go [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package cluster
```
Comment (LeonidVas) [inline cli/cluster/errors.go:1]: It doesn't seem to be related to "tt: initial support for cluster config".

**Review by LeonidVas — COMMENTED** [review]
Hi! Thank you for the patchset. That's great work.
In general, I have read the changes and have no objections. The only thing is that I have some wishes for the organization of commits in the future, to increase the atomicity of changes made by a single commit.

File: cli/running/script_instance.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
+package running
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/running/script_instance.go:1]: There is no point in creation an interface (Instance) with only one implementation - ScriptInstance.

File: cli/running/cluster_instance.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+		memtxDir:          instanceCtx.MemtxDir,
+		appDir:            instanceCtx.AppDir,
+		runDir:            instanceCtx.RunDir,
+		clusterConfigPath: instanceCtx.ClusterConfigPath,
+	}, nil
+}
+
+// shortenSocketPath reduces the length of console socket path.
+// It became common that console socket path is longer than 108/106 (on linux/macOs).
+func shortenSocketPath(socketPath string, basePath string) (string, error) {
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/running/cluster_instance.go:85]: This change is related to the commit. Script and cluster instances share same functionality to make console path shorter. So this code is extracted to the function as a part of this commit.

File: cli/running/cluster_instance.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+	// runDir is a directory that stores various instance runtime artifacts like
+	// console socket, PID file, etc.
+	runDir string
+	// clusterConfigPath is a path of the cluster config.
+	clusterConfigPath string
+}
+
+// NewInstance creates a clusterInstance.
+func newClusterInstance(tarantoolCli cmdcontext.TarantoolCli, instanceCtx InstanceCtx,
+	logger *ttlog.Logger) (
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/running/cluster_instance.go:53]: Fixed.

File: cli/cmd/start.go [hunk 9]
```diff
@@ -48,45 +49,64 @@ func NewStartCmd() *cobra.Command {
...
 
+	if canStart, reason :=
+		running.IsAbleToStartInstances(runningCtx.Instances, cmdCtx); !canStart {
+		return fmt.Errorf(reason)
+	}
+
 	if !watchdog {
-		ttBin, err := os.Executable()
-		if err != nil {
+		if err := startInstancesUnderWatchdog(runningCtx.Instances); err != nil {
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/cmd/start.go:107]: Done.

File: cli/running/cluster_instance.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+	"github.com/apex/log"
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/ttlog"
+	"github.com/tarantool/tt/cli/util"
+)
+
+// clusterInstance describes tarantool 3 instance running using cluster config.
+type clusterInstance struct {
+	processController *processController
```
[the remark this answers is not in the document]
Comment (psergee) [reply cli/running/cluster_instance.go:21]: Description is added.

File: cli/cluster/errors.go (hunk 5 above)
Comment (psergee) [reply cli/cluster/errors.go:1]: Sure. Done.

**Review by LeonidVas — APPROVED** [review]
